# Example of Kan-Derived Function fb

In [1]:
import os
from kan import *
import pandas as pd
from sklearn.model_selection import train_test_split, KFold
import numpy as np
import torch
import HydroErr as he

# Input

In [2]:
os.chdir('/Users/chuyangliu/Documents/Postdoc/Cive898/Writtings/GitHub_kan4baseflow')
os.getcwd()

'/Users/chuyangliu/Documents/Postdoc/Cive898/Writtings/GitHub_kan4baseflow'

In [3]:
col_index = 'Qb_hat/P_hat_true'
dir_in = r'./Datasets/Meira Neto et al 2020.xlsx' 

# Helper functions

In [4]:
def to_numpy_1d(array_like):
    import torch
    if isinstance(array_like, torch.Tensor):
        array_like = array_like.detach().cpu().numpy()
    return np.ravel(array_like)

def model_performance(sim_train, obs_train, sim_test, obs_test, additional_metrics=None):
    import HydroErr as he
    sim_train = to_numpy_1d(sim_train)
    obs_train = to_numpy_1d(obs_train)
    sim_test = to_numpy_1d(sim_test)
    obs_test = to_numpy_1d(obs_test)

    metrics = {
        'R^2': he.r_squared,
        'RMSE': he.rmse,
        'KGE': he.kge_2012,
        'NSE': he.nse
    }

    if additional_metrics:
        metrics.update(additional_metrics)

    def evaluate_metrics(sim, obs):
        results = {}
        for name, func in metrics.items():
            results[name] = func(sim, obs)
        return results

    train_results = evaluate_metrics(sim_train, obs_train)
    test_results = evaluate_metrics(sim_test, obs_test)

    print(f"Model performance on training data:")
    for name, value in train_results.items():
        print(f"{name} (train): {value}")

    print(f"\nModel performance on test data:")
    for name, value in test_results.items():
        print(f"{name} (test): {value}")


In [5]:
def average_r2(df_all, opt, lamb, lamb_coef, grid, seed, lib, x_label='Phi', y_label='Qb_hat/P_hat_true', n_splits=10,
               random_state=0, verbose=0):
    
    df_all['Qd_hat/P_hat_true'] = df_all['Qd'] / df_all['P']
    df_all['Qb_hat/P_hat_true'] = df_all['Qb'] / df_all['P']

    X_all = df_all[[x_label]].copy()
    y_all = df_all[y_label].values.reshape(-1, 1).copy()

    X, X_test_all, y, y_test_all = train_test_split(X_all, y_all, test_size=0.2, random_state=random_state)


    kf = KFold(n_splits=n_splits, shuffle=True, random_state=random_state)

    list_r2_test = []

    for fold, (train_index, test_index) in enumerate(kf.split(X)):
        print(f'\nFold {fold + 1}')
        X_train, X_test = X.iloc[train_index], X.iloc[test_index]
        y_train, y_test = y[train_index], y[test_index]

        n_repeat = 1
        X_train = np.repeat(X_train, n_repeat, axis=0)
        y_train = np.repeat(y_train, n_repeat, axis=0)
        X_test = X_test.values
        y_test = y_test.flatten()

        X_train = torch.from_numpy(X_train).float()
        X_test = torch.from_numpy(X_test).float()
        y_train = torch.from_numpy(y_train)
        y_test = torch.from_numpy(y_test)

        dataset = {
            'train_input': X_train,
            'test_input': X_test,
            'train_label': y_train,
            'test_label': y_test
        }

        model = KAN(width=[X_train.shape[1], 1], grid=grid, k=3, seed=seed)
        model(dataset['train_input'])

        model.train(dataset, opt=opt, steps=20, lamb=lamb, lamb_coef=lamb_coef)

        for epoch in range(2):
            results = model.train(dataset, opt=opt, steps=50, lamb=lamb, lamb_coef=lamb_coef)


        model.auto_symbolic(lib=lib)

        for epoch in range(1):
            model.train(dataset, opt=opt, steps=50, lamb=lamb, lamb_coef=lamb_coef)

        y_train_predicted = model(X_train)
        y_test_predicted = model(X_test)

        r2_train = he.r_squared(y_train_predicted.detach().numpy()[:, 0], y_train.detach().numpy().flatten())
        r2_test = he.r_squared(y_test_predicted.detach().numpy()[:, 0], y_test.detach().numpy().flatten())

        list_r2_test.append(r2_test)

        if verbose > 0:
            print(f"... KAN eqn: {model.symbolic_formula()[0][0]}")
            print(f'... Pre-training set (R^2): {r2_train}')
            print(f'... Validation set (R^2): {r2_test}')

        # -- all test
        y_pred_all = model(torch.from_numpy(X_test_all.values).float())
        r2_all = he.r_squared(y_pred_all.detach().numpy().flatten(), y_test_all.flatten())


        if verbose > 0:
            print(f"\n")
            model_performance(sim_train=model(torch.from_numpy(X.values).float()),
                              obs_train=y,
                              sim_test=y_pred_all,
                              obs_test=y_test_all,
                              additional_metrics=None)


    avg_r2 = np.nanmean(list_r2_test)


    return avg_r2

# Load Data


In [6]:
df_all = pd.read_excel(dir_in)

# Hyperparameter tuning

In [7]:
# Input
opt = "LBFGS"
lamb = 1e-2 
lamb_coef = 0.0
width = 1
lib = None

In [8]:
# -----   Grid search with 10-fold cross validation ----
# Example range of seeds and grid values to try
seed_range = range(0, 2)  
grid_range = range(2, 4)   

best_r2 = -float('inf')
best_seed = None
best_grid = None

# Iterate over different combinations of seed and grid
for seed in seed_range:
    for grid in grid_range:
        print(f"Testing seed={seed}, grid={grid}")

        avg_r2 = average_r2(df_all, opt, lamb, lamb_coef, grid, seed, lib,
                            x_label='Phi', y_label='Qb_hat/P_hat_true',
                            n_splits=10, random_state=5)

        print(f"Average R^2: {avg_r2}")

        # Update best parameters if current combination is better
        if avg_r2 > best_r2:
            best_r2 = avg_r2
            best_seed = seed
            best_grid = grid

print(f"\nBest parameters found:")
print(f"Seed: {best_seed}")
print(f"Grid: {best_grid}")
print(f"Best average R^2: {best_r2}")

Testing seed=0, grid=2

Fold 1


train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.10e-02 : 100%|██| 20/20 [00:00<00:00, 27.93it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.10e-02 : 100%|██| 50/50 [00:00<00:00, 96.71it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.10e-02 : 100%|██| 50/50 [00:00<00:00, 89.87it/s]


fixing (0,0,0) with x^4, r2=0.9998977780342102


train loss: 6.48e-02 | test loss: 1.38e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:00<00:00, 80.95it/s]



Fold 2


train loss: 5.90e-02 | test loss: 1.89e-01 | reg: 1.83e-02 : 100%|██| 20/20 [00:00<00:00, 61.34it/s]
train loss: 5.88e-02 | test loss: 3.46e+00 | reg: 1.87e-02 : 100%|██| 50/50 [00:01<00:00, 47.24it/s]
train loss: 5.88e-02 | test loss: 3.46e+00 | reg: 1.86e-02 : 100%|█| 50/50 [00:00<00:00, 117.10it/s]


fixing (0,0,0) with 1/x^2, r2=0.9993380308151245


train loss: 8.11e-02 | test loss: 2.34e-01 | reg: 1.25e-02 : 100%|██| 50/50 [00:01<00:00, 43.02it/s]



Fold 3


train loss: 6.19e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 61.84it/s]
train loss: 6.19e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 131.52it/s]
train loss: 6.19e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 160.67it/s]


fixing (0,0,0) with x^4, r2=0.9999616742134094


train loss: 6.49e-02 | test loss: 1.36e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 91.18it/s]



Fold 4


train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 44.75it/s]
train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 165.73it/s]
train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 155.10it/s]


fixing (0,0,0) with cosh, r2=0.9996088743209839


train loss: nan | test loss: nan | reg: nan : 100%|█████████████████| 50/50 [00:01<00:00, 26.76it/s]
/Users/chuyangliu/anaconda3/envs/cive898/lib/python3.10/site-packages/HydroErr/HydroErr.py:6248: UserWarning: Row(s) [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192


Fold 5


train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 39.18it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:01<00:00, 43.29it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:01<00:00, 39.56it/s]


fixing (0,0,0) with arctan, r2=0.9992539286613464


train loss: 5.77e-02 | test loss: 1.58e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:01<00:00, 47.19it/s]



Fold 6


train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.08e-02 : 100%|██| 20/20 [00:00<00:00, 35.08it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:01<00:00, 35.94it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:01<00:00, 39.59it/s]


fixing (0,0,0) with arctan, r2=0.9991993308067322


train loss: 6.09e-02 | test loss: 1.70e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 145.88it/s]



Fold 7


train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.08e-02 : 100%|██| 20/20 [00:00<00:00, 41.62it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.08e-02 : 100%|█| 50/50 [00:00<00:00, 128.36it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.08e-02 : 100%|█| 50/50 [00:00<00:00, 138.83it/s]


fixing (0,0,0) with exp, r2=0.9992741942405701


train loss: 6.19e-02 | test loss: 1.57e-01 | reg: 1.06e-02 : 100%|█| 50/50 [00:00<00:00, 139.07it/s]



Fold 8


train loss: 6.06e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 20/20 [00:00<00:00, 30.82it/s]
train loss: 6.02e-02 | test loss: 1.27e-01 | reg: 1.12e-02 : 100%|██| 50/50 [00:00<00:00, 65.79it/s]
train loss: 6.02e-02 | test loss: 1.27e-01 | reg: 1.12e-02 : 100%|█| 50/50 [00:00<00:00, 115.26it/s]


fixing (0,0,0) with exp, r2=0.9994964003562927


train loss: 6.08e-02 | test loss: 1.25e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 118.12it/s]



Fold 9


train loss: 5.93e-02 | test loss: 1.69e-01 | reg: 1.08e-02 : 100%|██| 20/20 [00:00<00:00, 30.69it/s]
train loss: 5.92e-02 | test loss: 1.69e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 70.60it/s]
train loss: 5.92e-02 | test loss: 1.69e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:00<00:00, 61.98it/s]


fixing (0,0,0) with cosh, r2=0.9997639060020447


train loss: 5.87e-02 | test loss: 1.70e-01 | reg: 1.05e-02 : 100%|██| 50/50 [00:00<00:00, 71.51it/s]



Fold 10


train loss: 5.97e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|██| 20/20 [00:00<00:00, 47.38it/s]
train loss: 5.97e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 178.56it/s]
train loss: 5.97e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 158.97it/s]


fixing (0,0,0) with x^4, r2=0.9999455809593201


train loss: 6.24e-02 | test loss: 1.60e-01 | reg: 1.34e-02 : 100%|██| 50/50 [00:00<00:00, 86.55it/s]


Average R^2: 0.6950700686614179
Testing seed=0, grid=3

Fold 1


train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 38.43it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 117.43it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 103.34it/s]


fixing (0,0,0) with x^3, r2=0.9999996423721313


train loss: 7.01e-02 | test loss: 1.37e-01 | reg: 1.01e-02 : 100%|██| 50/50 [00:00<00:00, 92.79it/s]



Fold 2


train loss: 5.91e-02 | test loss: 1.84e-01 | reg: 1.84e-02 : 100%|██| 20/20 [00:00<00:00, 57.19it/s]
train loss: 5.91e-02 | test loss: 1.84e-01 | reg: 1.84e-02 : 100%|██| 50/50 [00:00<00:00, 72.27it/s]
train loss: 5.91e-02 | test loss: 1.84e-01 | reg: 1.84e-02 : 100%|██| 50/50 [00:00<00:00, 81.45it/s]


fixing (0,0,0) with cosh, r2=0.9870108366012573


train loss: 5.95e-02 | test loss: 1.86e-01 | reg: 1.79e-02 : 100%|██| 50/50 [00:00<00:00, 94.18it/s]



Fold 3


train loss: 6.20e-02 | test loss: 1.43e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 57.29it/s]
train loss: 6.20e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 68.21it/s]
train loss: 6.20e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 135.92it/s]


fixing (0,0,0) with x^4, r2=0.9999532699584961


train loss: 6.49e-02 | test loss: 1.36e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 101.25it/s]



Fold 4


train loss: 6.17e-02 | test loss: 1.29e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 43.67it/s]
train loss: 6.17e-02 | test loss: 1.29e-01 | reg: 1.12e-02 : 100%|██| 50/50 [00:00<00:00, 75.10it/s]
train loss: 6.17e-02 | test loss: 1.29e-01 | reg: 1.12e-02 : 100%|█| 50/50 [00:00<00:00, 134.84it/s]


fixing (0,0,0) with tanh, r2=0.9992700219154358


train loss: 6.14e-02 | test loss: 1.29e-01 | reg: 1.15e-02 : 100%|█| 50/50 [00:00<00:00, 125.62it/s]



Fold 5


train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 54.42it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.12e-02 : 100%|█| 50/50 [00:00<00:00, 106.08it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 72.07it/s]


fixing (0,0,0) with x^4, r2=0.9991799592971802


train loss: 6.17e-02 | test loss: 1.62e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:00<00:00, 91.15it/s]



Fold 6


train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|██| 20/20 [00:00<00:00, 31.58it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 80.90it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 100.80it/s]


fixing (0,0,0) with arctan, r2=0.998528778553009


train loss: 6.09e-02 | test loss: 1.70e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 100.47it/s]



Fold 7


train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 20/20 [00:00<00:00, 38.71it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 64.87it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 94.73it/s]


fixing (0,0,0) with tanh, r2=0.9990874528884888


train loss: 6.13e-02 | test loss: 1.58e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 134.67it/s]



Fold 8


train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 20/20 [00:00<00:00, 41.35it/s]
train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:00<00:00, 86.71it/s]
train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:00<00:00, 90.68it/s]


fixing (0,0,0) with arctan, r2=0.9981390833854675


train loss: 6.01e-02 | test loss: 1.27e-01 | reg: 1.14e-02 : 100%|█| 50/50 [00:00<00:00, 118.04it/s]



Fold 9


train loss: 5.82e-02 | test loss: 1.69e-01 | reg: 1.08e-02 : 100%|██| 20/20 [00:00<00:00, 53.60it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|██| 50/50 [00:00<00:00, 55.66it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|█| 50/50 [00:00<00:00, 135.08it/s]


fixing (0,0,0) with x^4, r2=0.9999510049819946


train loss: 6.15e-02 | test loss: 1.67e-01 | reg: 1.05e-02 : 100%|██| 50/50 [00:00<00:00, 91.17it/s]



Fold 10


train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|██| 20/20 [00:00<00:00, 36.85it/s]
train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 146.64it/s]
train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|██| 50/50 [00:00<00:00, 90.25it/s]


fixing (0,0,0) with x^3, r2=0.9999986886978149


train loss: 1.10e-01 | test loss: 1.23e-01 | reg: 9.72e-03 : 100%|█| 50/50 [00:00<00:00, 118.26it/s]


Average R^2: 0.7454046508549044
Testing seed=1, grid=2

Fold 1


train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.10e-02 : 100%|██| 20/20 [00:00<00:00, 47.58it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 122.09it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 103.79it/s]


fixing (0,0,0) with x^4, r2=0.999898374080658


train loss: 6.48e-02 | test loss: 1.38e-01 | reg: 1.46e-02 : 100%|█| 50/50 [00:00<00:00, 143.09it/s]



Fold 2


train loss: 5.90e-02 | test loss: 1.88e-01 | reg: 1.83e-02 : 100%|██| 20/20 [00:00<00:00, 67.86it/s]
train loss: 5.90e-02 | test loss: 1.88e-01 | reg: 1.83e-02 : 100%|██| 50/50 [00:00<00:00, 80.48it/s]
train loss: 5.90e-02 | test loss: 1.88e-01 | reg: 1.83e-02 : 100%|██| 50/50 [00:00<00:00, 59.72it/s]


fixing (0,0,0) with arctan, r2=0.9985855221748352


train loss: 5.89e-02 | test loss: 1.89e-01 | reg: 1.86e-02 : 100%|██| 50/50 [00:00<00:00, 87.39it/s]



Fold 3


train loss: 6.19e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 38.09it/s]
train loss: 6.19e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 118.13it/s]
train loss: 6.19e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 135.13it/s]


fixing (0,0,0) with x^4, r2=0.9999620914459229


train loss: 6.49e-02 | test loss: 1.36e-01 | reg: 1.52e-02 : 100%|█| 50/50 [00:00<00:00, 131.71it/s]



Fold 4


train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 56.34it/s]
train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 131.63it/s]
train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 122.17it/s]


fixing (0,0,0) with cosh, r2=0.9996089935302734


train loss: 6.21e-02 | test loss: 1.27e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 88.60it/s]



Fold 5


train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 47.58it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 136.82it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 152.20it/s]


fixing (0,0,0) with arctan, r2=0.9992537498474121


train loss: 5.77e-02 | test loss: 1.58e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:00<00:00, 91.40it/s]



Fold 6


train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.08e-02 : 100%|██| 20/20 [00:00<00:00, 31.47it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.08e-02 : 100%|█| 50/50 [00:00<00:00, 115.08it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.08e-02 : 100%|█| 50/50 [00:00<00:00, 109.36it/s]


fixing (0,0,0) with arctan, r2=0.9991776943206787


train loss: 6.09e-02 | test loss: 1.70e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 149.02it/s]



Fold 7


train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 20/20 [00:00<00:00, 30.21it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:00<00:00, 80.30it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.08e-02 : 100%|█| 50/50 [00:00<00:00, 165.63it/s]


fixing (0,0,0) with exp, r2=0.999275267124176


train loss: 6.19e-02 | test loss: 1.57e-01 | reg: 1.06e-02 : 100%|█| 50/50 [00:00<00:00, 134.11it/s]



Fold 8


train loss: 6.02e-02 | test loss: 1.27e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 50.48it/s]
train loss: 6.02e-02 | test loss: 1.27e-01 | reg: 1.12e-02 : 100%|█| 50/50 [00:00<00:00, 162.11it/s]
train loss: 6.02e-02 | test loss: 1.27e-01 | reg: 1.12e-02 : 100%|█| 50/50 [00:00<00:00, 115.34it/s]


fixing (0,0,0) with exp, r2=0.999496340751648


train loss: 6.06e-02 | test loss: 1.26e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 146.84it/s]



Fold 9


train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|██| 20/20 [00:00<00:00, 55.92it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|█| 50/50 [00:00<00:00, 167.76it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|█| 50/50 [00:00<00:00, 126.77it/s]


fixing (0,0,0) with x^4, r2=0.9998205900192261


train loss: 6.15e-02 | test loss: 1.67e-01 | reg: 1.05e-02 : 100%|█| 50/50 [00:00<00:00, 105.72it/s]



Fold 10


train loss: 5.97e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|██| 20/20 [00:00<00:00, 49.37it/s]
train loss: 5.97e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 169.47it/s]
train loss: 5.97e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 141.28it/s]


fixing (0,0,0) with x^4, r2=0.9999457597732544


train loss: 6.24e-02 | test loss: 1.61e-01 | reg: 1.07e-02 : 100%|█| 50/50 [00:00<00:00, 127.85it/s]


Average R^2: 0.7396752785197427
Testing seed=1, grid=3

Fold 1


train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 61.15it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 76.80it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 116.41it/s]


fixing (0,0,0) with x^3, r2=0.9999993443489075


train loss: 7.01e-02 | test loss: 1.37e-01 | reg: 1.01e-02 : 100%|██| 50/50 [00:00<00:00, 85.47it/s]



Fold 2


train loss: 5.91e-02 | test loss: 5.47e-01 | reg: 1.84e-02 : 100%|██| 20/20 [00:00<00:00, 48.89it/s]
train loss: 5.91e-02 | test loss: 5.49e-01 | reg: 1.84e-02 : 100%|██| 50/50 [00:00<00:00, 82.49it/s]
train loss: 5.91e-02 | test loss: 5.49e-01 | reg: 1.84e-02 : 100%|█| 50/50 [00:00<00:00, 113.45it/s]


fixing (0,0,0) with cosh, r2=0.9985430836677551


train loss: 5.95e-02 | test loss: 2.57e-01 | reg: 1.79e-02 : 100%|██| 50/50 [00:00<00:00, 70.31it/s]



Fold 3


train loss: 6.20e-02 | test loss: 1.43e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 61.46it/s]
train loss: 6.20e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 75.33it/s]
train loss: 6.20e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 88.95it/s]


fixing (0,0,0) with x^4, r2=0.9999778866767883


train loss: 6.49e-02 | test loss: 1.36e-01 | reg: 1.51e-02 : 100%|█| 50/50 [00:00<00:00, 134.88it/s]



Fold 4


train loss: 6.17e-02 | test loss: 1.29e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 64.59it/s]
train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.12e-02 : 100%|██| 50/50 [00:00<00:00, 76.90it/s]
train loss: 6.17e-02 | test loss: 1.28e-01 | reg: 1.12e-02 : 100%|██| 50/50 [00:00<00:00, 89.35it/s]


fixing (0,0,0) with tanh, r2=0.9993158578872681


train loss: 6.14e-02 | test loss: 1.29e-01 | reg: 1.15e-02 : 100%|██| 50/50 [00:00<00:00, 79.63it/s]



Fold 5


train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 52.93it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 82.56it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 94.74it/s]


fixing (0,0,0) with tanh, r2=0.9991652369499207


train loss: 5.78e-02 | test loss: 1.59e-01 | reg: 1.14e-02 : 100%|█| 50/50 [00:00<00:00, 120.26it/s]



Fold 6


train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|██| 20/20 [00:00<00:00, 44.24it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 95.27it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 108.09it/s]


fixing (0,0,0) with arctan, r2=0.9985214471817017


train loss: 6.09e-02 | test loss: 1.70e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 151.81it/s]



Fold 7


train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 20/20 [00:00<00:00, 62.04it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 73.48it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 61.36it/s]


fixing (0,0,0) with tanh, r2=0.9991387724876404


train loss: 6.13e-02 | test loss: 1.58e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 146.30it/s]



Fold 8


train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 20/20 [00:00<00:00, 60.32it/s]
train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:00<00:00, 89.66it/s]
train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:00<00:00, 99.39it/s]


fixing (0,0,0) with arctan, r2=0.9981462359428406


train loss: 6.01e-02 | test loss: 1.27e-01 | reg: 1.14e-02 : 100%|█| 50/50 [00:00<00:00, 106.30it/s]



Fold 9


train loss: 5.82e-02 | test loss: 1.69e-01 | reg: 1.08e-02 : 100%|██| 20/20 [00:00<00:00, 56.30it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:00<00:00, 73.40it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|██| 50/50 [00:00<00:00, 78.04it/s]


fixing (0,0,0) with x^4, r2=0.9999597668647766


train loss: 6.15e-02 | test loss: 1.67e-01 | reg: 1.05e-02 : 100%|██| 50/50 [00:00<00:00, 63.61it/s]



Fold 10


train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|██| 20/20 [00:00<00:00, 66.95it/s]
train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 118.78it/s]
train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 103.89it/s]


fixing (0,0,0) with x^3, r2=0.9999986290931702


train loss: 6.76e-02 | test loss: 1.53e-01 | reg: 9.95e-03 : 100%|█| 50/50 [00:00<00:00, 141.65it/s]

Average R^2: 0.6671218566198217

Best parameters found:
Seed: 0
Grid: 3
Best average R^2: 0.7454046508549044


# Check the optimal model with the best parameters

In [9]:
average_r2(df_all, opt, lamb, lamb_coef, best_grid, best_seed, lib,
           x_label='Phi', y_label='Qb_hat/P_hat_true',
           n_splits=10, random_state=5, verbose = 2)


Fold 1


train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|██| 20/20 [00:00<00:00, 40.01it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 130.11it/s]
train loss: 6.16e-02 | test loss: 1.34e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 110.45it/s]


fixing (0,0,0) with x^3, r2=0.9999996423721313


train loss: 7.01e-02 | test loss: 1.37e-01 | reg: 1.01e-02 : 100%|█| 50/50 [00:00<00:00, 101.57it/s]


... KAN eqn: 0.52*(1 - 0.24*x_1)**3 - 0.e-2
... Pre-training set (R^2): 0.6639028293563374
... Validation set (R^2): 0.7493244706449997


Model performance on training data:
R^2 (train): 0.6706621257799479
RMSE (train): 0.0686342646624862
KGE (train): 0.7371960596143217
NSE (train): 0.670584567585866

Model performance on test data:
R^2 (test): 0.6590677133713186
RMSE (test): 0.06775534440622558
KGE (test): 0.6887054591594054
NSE (test): 0.6515954861586928

Fold 2


train loss: 5.91e-02 | test loss: 1.84e-01 | reg: 1.84e-02 : 100%|██| 20/20 [00:00<00:00, 60.95it/s]
train loss: 5.91e-02 | test loss: 1.84e-01 | reg: 1.84e-02 : 100%|██| 50/50 [00:00<00:00, 78.68it/s]
train loss: 5.91e-02 | test loss: 1.84e-01 | reg: 1.84e-02 : 100%|██| 50/50 [00:00<00:00, 87.81it/s]


fixing (0,0,0) with cosh, r2=0.9870108366012573


train loss: 5.95e-02 | test loss: 1.86e-01 | reg: 1.79e-02 : 100%|██| 50/50 [00:00<00:00, 99.29it/s]


... KAN eqn: 0.03
... Pre-training set (R^2): 0.7399408786286475
... Validation set (R^2): 0.7593354038924806


Model performance on training data:
R^2 (train): 0.742300975031701
RMSE (train): 0.060724108480515605
KGE (train): 0.7996667360335922
NSE (train): 0.7421398293055952

Model performance on test data:
R^2 (test): 0.7114599897626578
RMSE (test): 0.062093930376120085
KGE (test): 0.7809969407145749
NSE (test): 0.7073860953839091

Fold 3


train loss: 6.20e-02 | test loss: 1.43e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 60.10it/s]
train loss: 6.20e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 73.19it/s]
train loss: 6.20e-02 | test loss: 1.42e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 161.50it/s]


fixing (0,0,0) with x^4, r2=0.9999532699584961


train loss: 6.49e-02 | test loss: 1.36e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 114.73it/s]


... KAN eqn: 0.62*(1 - 0.24*x_1)**4 + 0.01
... Pre-training set (R^2): 0.7139496184057499
... Validation set (R^2): 0.847825247541654


Model performance on training data:
R^2 (train): 0.7189859293341181
RMSE (train): 0.06345479980989682
KGE (train): 0.770833543773507
NSE (train): 0.7184270521632046

Model performance on test data:
R^2 (test): 0.6998565987076784
RMSE (test): 0.06368444179134618
KGE (test): 0.7491158343685655
NSE (test): 0.6922037307092512

Fold 4


train loss: 6.17e-02 | test loss: 1.29e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 47.76it/s]
train loss: 6.17e-02 | test loss: 1.29e-01 | reg: 1.12e-02 : 100%|██| 50/50 [00:00<00:00, 79.79it/s]
train loss: 6.17e-02 | test loss: 1.29e-01 | reg: 1.12e-02 : 100%|█| 50/50 [00:00<00:00, 138.97it/s]


fixing (0,0,0) with tanh, r2=0.9992700219154358


train loss: 6.14e-02 | test loss: 1.29e-01 | reg: 1.15e-02 : 100%|█| 50/50 [00:00<00:00, 138.45it/s]


... KAN eqn: 0.39 - 0.34*tanh(1.42*x_1 - 0.82)
... Pre-training set (R^2): 0.7467385645332453
... Validation set (R^2): 0.7722991462248904


Model performance on training data:
R^2 (train): 0.7474905810222926
RMSE (train): 0.06010343033108799
KGE (train): 0.8040276374351344
NSE (train): 0.7473842117724503

Model performance on test data:
R^2 (test): 0.7440847811189126
RMSE (test): 0.058473655281330465
KGE (test): 0.7943953652440818
NSE (test): 0.7405120799817747

Fold 5


train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.12e-02 : 100%|██| 20/20 [00:00<00:00, 57.41it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.12e-02 : 100%|█| 50/50 [00:00<00:00, 115.44it/s]
train loss: 5.81e-02 | test loss: 1.61e-01 | reg: 1.11e-02 : 100%|██| 50/50 [00:00<00:00, 74.95it/s]


fixing (0,0,0) with x^4, r2=0.9991799592971802


train loss: 6.17e-02 | test loss: 1.62e-01 | reg: 1.08e-02 : 100%|██| 50/50 [00:00<00:00, 93.70it/s]


... KAN eqn: 0.63*(1 - 0.24*x_1)**4 + 0.01
... Pre-training set (R^2): 0.7303405452633105
... Validation set (R^2): 0.6251395347208688


Model performance on training data:
R^2 (train): 0.7190430674604805
RMSE (train): 0.06338651285881212
KGE (train): 0.7849563452241733
NSE (train): 0.7190327561347556

Model performance on test data:
R^2 (test): 0.7004642164145592
RMSE (test): 0.06325701233115784
KGE (test): 0.7665780819234923
NSE (test): 0.696321524360779

Fold 6


train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|██| 20/20 [00:00<00:00, 30.36it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 82.62it/s]
train loss: 6.10e-02 | test loss: 1.71e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 112.32it/s]


fixing (0,0,0) with arctan, r2=0.998528778553009


train loss: 6.09e-02 | test loss: 1.70e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 115.69it/s]


... KAN eqn: 0.35 - 0.23*atan(2.34*x_1 - 1.53)
... Pre-training set (R^2): 0.7351202699912679
... Validation set (R^2): 0.8375636221870653


Model performance on training data:
R^2 (train): 0.7473431454967762
RMSE (train): 0.06012164229720497
KGE (train): 0.8006136997666389
NSE (train): 0.7472310981424986

Model performance on test data:
R^2 (test): 0.7432412318976168
RMSE (test): 0.058454061339499236
KGE (test): 0.7837769824662804
NSE (test): 0.740685954500881

Fold 7


train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 20/20 [00:00<00:00, 54.38it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|██| 50/50 [00:00<00:00, 68.97it/s]
train loss: 6.15e-02 | test loss: 1.57e-01 | reg: 1.09e-02 : 100%|█| 50/50 [00:00<00:00, 103.20it/s]


fixing (0,0,0) with tanh, r2=0.9990874528884888


train loss: 6.13e-02 | test loss: 1.58e-01 | reg: 1.11e-02 : 100%|█| 50/50 [00:00<00:00, 142.47it/s]


... KAN eqn: 0.43 - 0.38*tanh(1.34*x_1 - 0.67)
... Pre-training set (R^2): 0.7393297837804095
... Validation set (R^2): 0.8400163419783512


Model performance on training data:
R^2 (train): 0.7474610180748398
RMSE (train): 0.060111034113169905
KGE (train): 0.8030136268524994
NSE (train): 0.7473202900662332

Model performance on test data:
R^2 (test): 0.7398998351471184
RMSE (test): 0.059004565917539976
KGE (test): 0.7920622236045071
NSE (test): 0.7357786555652777

Fold 8


train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 20/20 [00:00<00:00, 42.17it/s]
train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:00<00:00, 87.53it/s]
train loss: 6.02e-02 | test loss: 1.28e-01 | reg: 1.13e-02 : 100%|██| 50/50 [00:00<00:00, 92.08it/s]


fixing (0,0,0) with arctan, r2=0.9981390833854675


train loss: 6.01e-02 | test loss: 1.27e-01 | reg: 1.14e-02 : 100%|█| 50/50 [00:00<00:00, 133.52it/s]


... KAN eqn: 0.35 - 0.22*atan(2.31*x_1 - 1.55)
... Pre-training set (R^2): 0.7546117871241649
... Validation set (R^2): 0.6282923707483421


Model performance on training data:
R^2 (train): 0.7474647307506397
RMSE (train): 0.06009557929722565
KGE (train): 0.8080155041617537
NSE (train): 0.7474502035323969

Model performance on test data:
R^2 (test): 0.7450739778346601
RMSE (test): 0.05820367680132587
KGE (test): 0.7920149900225495
NSE (test): 0.7429027096044658

Fold 9


train loss: 5.82e-02 | test loss: 1.69e-01 | reg: 1.08e-02 : 100%|██| 20/20 [00:00<00:00, 58.30it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|██| 50/50 [00:00<00:00, 56.79it/s]
train loss: 5.82e-02 | test loss: 1.70e-01 | reg: 1.07e-02 : 100%|█| 50/50 [00:00<00:00, 132.19it/s]


fixing (0,0,0) with x^4, r2=0.9999510049819946


train loss: 6.15e-02 | test loss: 1.67e-01 | reg: 1.05e-02 : 100%|██| 50/50 [00:00<00:00, 91.61it/s]


... KAN eqn: 0.62*(1 - 0.24*x_1)**4 + 0.01
... Pre-training set (R^2): 0.7231243583021
... Validation set (R^2): 0.6893421521892057


Model performance on training data:
R^2 (train): 0.719036825627427
RMSE (train): 0.06351083505313247
KGE (train): 0.7829546143082113
NSE (train): 0.717929533449492

Model performance on test data:
R^2 (test): 0.7003562641141129
RMSE (test): 0.06295778598283883
KGE (test): 0.7671710875907053
NSE (test): 0.6991877261901824

Fold 10


train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|██| 20/20 [00:00<00:00, 37.74it/s]
train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|█| 50/50 [00:00<00:00, 149.87it/s]
train loss: 5.96e-02 | test loss: 1.65e-01 | reg: 1.10e-02 : 100%|██| 50/50 [00:00<00:00, 94.90it/s]


fixing (0,0,0) with x^3, r2=0.9999986886978149


train loss: 1.10e-01 | test loss: 1.23e-01 | reg: 9.72e-03 : 100%|█| 50/50 [00:00<00:00, 114.37it/s]


... KAN eqn: 0.11*(1 - 0.4*x_1)**3 + 0.17
... Pre-training set (R^2): 0.14571679253217518
... Validation set (R^2): 0.7049082184211856


Model performance on training data:
R^2 (train): 0.15314886347845938
RMSE (train): 0.11009045973432595
KGE (train): 0.12055943591271734
NSE (train): 0.1524577800824073

Model performance on test data:
R^2 (test): 0.6915225088015506
RMSE (test): 0.10254940418025006
KGE (test): 0.11276289356104474
NSE (test): 0.20188965248469604


0.7454046508549044

# Optimal model and its model performance
'''
... KAN eqn: 0.39 - 0.34*tanh(1.42*x_1 - 0.82)

Model performance on training data:
R^2 (train): 0.7474905810222926
RMSE (train): 0.06010343033108799
KGE (train): 0.8040276374351344
NSE (train): 0.7473842117724503

Model performance on test data:
R^2 (test): 0.7440847811189126
RMSE (test): 0.058473655281330465
KGE (test): 0.7943953652440818
NSE (test): 0.7405120799817747
'''